### PR TITLE
Allow users to load static landing page from config/initializers

### PR DESCRIPTION
**Pull request by madumlao**
By default, if clp_static_enabled: true, then landing page data is read
from app/services/landing_page/example_data.rb. The docs then recommend
editing that file directly to specify the landing page structure.

This is error prone, git-unfriendly, and un-rails-like. The more railsy
approach would be to specify the landing page as configuration data but
default to example_data.rb when the config data is absent.

This commit does just that. If clp_static_enabled, then landing page
data is loaded from CustomLandingPage::StaticData, but if that is not
present, will fall back to loading CustomLandingPage::ExampleData.

CustomLandingPage::StaticData can be populated from a file in
config/initializers. This initializer can automatically be generated
through a rake task, i.e., `sharetribe:landing_pages:install_static`.
The landing page can now be edited from the generated file.

### CODE REVIEW

File: app/services/custom_landing_page/landing_page.rb.template [hunk 1]
```diff
@@ -0,0 +1,304 @@
+module CustomLandingPage
```
Comment (valdis) [inline app/services/custom_landing_page/landing_page.rb.template]: This duplicates `app/services/custom_landing_page/example_data.rb`. can you reuse?

File: app/controllers/landing_page_controller.rb [hunk 2]
```diff
@@ -123,8 +123,14 @@ def preview
     begin
       structure = CLP::LandingPageStore.load_structure(cid, preview_version)
 
-      # Uncomment to use static data instead of dynamic from DB
-      # structure = JSON.parse(CustomLandingPage::ExampleData::DATA_STR)
+      # if static landing page enabled, load data from config/initializers/landing_page.rb
+      if (APP_CONFIG.clp_static_enabled)
+        begin
```
Comment (valdis) [inline app/controllers/landing_page_controller.rb]: don't use exceptions for flow control

File: app/services/custom_landing_page/landing_page_store_static.rb [hunk 3]
```diff
@@ -10,8 +10,15 @@ def released_version(*)
     end
 
     def load_structure(*)
-      LandingPageStoreDefaults.add_defaults(
+      structure = LandingPageStoreDefaults.add_defaults(
         JSON.parse(CustomLandingPage::ExampleData::DATA_STR))
+      
+      begin
```
Comment (valdis) [inline app/services/custom_landing_page/landing_page_store_static.rb]: don't use exceptions for flow control

File: lib/tasks/sharetribe.rake [hunk 4]
```diff
@@ -7,6 +7,15 @@ namespace :sharetribe do
     end
   end
 
+  namespace :landing_pages do
+    desc "Install sample landing page into initializers/landing_page.rb"
+    task :install_static => :environment do
+      source = File.join(Rails.root, "app", "services", "custom_landing_page", "landing_page.rb.template")
```
Comment (valdis) [inline lib/tasks/sharetribe.rake]: a newcomer to ShareTribe OS might be confused why static landing pages don't work after running this rake task. Some output to remind setting `clp_static_enabled: true` might be helpful

Comment (madumlao) [reply lib/tasks/sharetribe.rake]: Added note in ec4170606b687f1e0bccc1bb33a3b9e89a7f7274

File: app/controllers/landing_page_controller.rb (hunk 2 above)
Comment (madumlao) [reply app/controllers/landing_page_controller.rb]: Changed flow control to use Module#const_defined? in 2451599639851d7721dc02dfc78078e7bdcabdf9

File: app/services/custom_landing_page/landing_page_store_static.rb (hunk 3 above)
Comment (madumlao) [reply app/services/custom_landing_page/landing_page_store_static.rb]: Changed flow control to use Module#const_defined? in 2451599639851d7721dc02dfc78078e7bdcabdf9

File: app/services/custom_landing_page/landing_page.rb.template (hunk 1 above)
Comment (madumlao) [reply app/services/custom_landing_page/landing_page.rb.template]: Could you give suggestions on how I can reuse here? I duplicated some but not all of `ExampleData` and the module name differs by a bit.

Would a copy + regex substitution suffice?

File: app/controllers/landing_page_controller.rb [hunk 5]
```diff
@@ -123,8 +123,14 @@ def preview
     begin
       structure = CLP::LandingPageStore.load_structure(cid, preview_version)
 
-      # Uncomment to use static data instead of dynamic from DB
-      # structure = JSON.parse(CustomLandingPage::ExampleData::DATA_STR)
+      # if static landing page enabled, load data from config/initializers/landing_page.rb
+      if (APP_CONFIG.clp_static_enabled)
```
Comment (valdis) [inline app/controllers/landing_page_controller.rb]: this whole block seems unnecessary, structure is set to the correct value already in `app/services/custom_landing_page/landing_page_store_static.rb` when you call `load_structure`

IMHO you can just update the comment section to say that one should run the rake task and set clp_static_enabled: true

File: app/services/custom_landing_page/landing_page.rb.template (hunk 1 above)
Comment (valdis) [reply app/services/custom_landing_page/landing_page.rb.template]: 1) Stick to `example_data.rb` file. 
2) You can keep your version of `DATA_STR`, just set the `copyright` text back to "This website is powered by Sharetribe marketplace platform."
3) in the rake task copy the `example_data.rb` and replace the module name to `StaticData` in the new file
4) remove the landing_page.rb.template

File: app/services/custom_landing_page/landing_page_store_static.rb [hunk 6]
```diff
@@ -10,8 +10,14 @@ def released_version(*)
...
     def load_structure(*)
-      LandingPageStoreDefaults.add_defaults(
+      structure = LandingPageStoreDefaults.add_defaults(
         JSON.parse(CustomLandingPage::ExampleData::DATA_STR))
+      
+      if CustomLandingPage.const_defined?("StaticData")
+        structure = JSON.parse(CustomLandingPage::StaticData::DATA_STR)
+      end
+
+      structure
```
Comment (valdis) [inline app/services/custom_landing_page/landing_page_store_static.rb]: how about
```
def load_structure(*)
  data = if CustomLandingPage.const_defined?("StaticData")
           CustomLandingPage::StaticData::DATA_STR
         else
           CustomLandingPage::ExampleData::DATA_STR
         end
  LandingPageStoreDefaults.add_defaults((JSON.parse(data))
end
```

**Comment by madumlao** [comment]
Updated rake task to pull / patch from CustomLandingPage::ExampleData

Simplified logic of choosing landing pages according to suggestions

**Comment by valdis** [comment]
Good stuff! Thanks!

**Comment by aivils** [comment]
Opened fresh PR https://github.com/sharetribe/sharetribe/pull/3316 . This is closed.

**Pull request closed without merging**